src/background/EventManager.ts

```diff
@@ -46,37 +28,47 @@ export class EventManager {
       types: ["main_frame"],
     };
     const extraInfoSpec: WebRequest.OnHeadersReceivedOptions[] = ["blocking"];
-    this.webRequestHeadersEmitter.addListener(
-      this.receiveWebRequest.bind(this),
-      filter,
-      extraInfoSpec
-    );
-    this.webRequestErrorEmitter.addListener(
-      this.receiveWebRequestError.bind(this)
-    );
-    this.messageEmitter.addListener(this.receiveMessage.bind(this));
-    this.tabActivatedEmitter.addListener(this.changeBrowserAction.bind(this));
-    this.webRequestBeforeEmitter.addListener(
+
+    this.webRequest.onBeforeRequest.addListener(
       this.resetTabData.bind(this),
       filter,
       []
     );
+    this.webRequest.onHeadersReceived.addListener(
+      this.receiveWebRequestHeaders.bind(this),
+      filter,
+      extraInfoSpec
+    );
+    this.webNavigation.onErrorOccurred.addListener(
+      this.receiveWebNavigationError.bind(this)
+    );
+    this.runtime.onMessage.addListener(this.receiveMessage.bind(this));
+    this.tabs.onActivated.addListener(this.changeBrowserAction.bind(this));
   }
 
   public resetTabData(requestDetails: { tabId: number }): void {
     this.app.resetTabData(requestDetails.tabId);
   }
 
-  public receiveWebRequest(
+  public async receiveWebRequestHeaders(
     requestDetails: WebRequest.OnHeadersReceivedDetailsType
-  ): void {
-    // using await is not possible here, since making receiveWebRequest async is not allowed
-    this.app.fetchCertificate(requestDetails).then(() => {
-      this.changeBrowserAction(requestDetails);
-    });
+  ): Promise<WebRequest.BlockingResponse> {
+    await this.app.fetchCertificate(requestDetails);
+    const hasQualityDecreased = await this.app.analyzeQuality(requestDetails);

```

Comment:
Has to be documented well since the method name does not indicate that a boolean regarding quality decrease is returned.